consensus/istanbul/backend/engine.go

```diff
@@ -390,6 +425,27 @@ func (sb *Backend) Prepare(chain consensus.ChainReader, header *types.Header) er
 	// wait for the timestamp of header, use this to adjust the block period
 	delay := time.Unix(header.Time.Int64(), 0).Sub(now())
 	time.Sleep(delay)
+
+	// modify the block header to include all the ParentCommits
+	// only do this for blocks which start with block 1 as a parent
+	if header.Number.Uint64() > 1 {
+		// copy over the seals we have saved the previous block
+		parentExtra, err := types.ExtractIstanbulExtra(parent)
+		if err != nil {
+			return err
+		}
+		bitmap := parentExtra.Bitmap
+		asig := parentExtra.CommittedSeal
+
+		parentGossipedSeals := sb.core.ParentCommits()
+		// if we had any seals gossiped to us, proceed to add them to the
+		// already aggregated signature
+		if parentGossipedSeals != nil && parentGossipedSeals.Size() != 0 {
+			bitmap, asig = istanbulCore.UnionOfSeals(bitmap, asig, parentGossipedSeals)
```

Comment:
Just to be clear, @kevjue we believe that there is no risk that `parentCommits` will have messages for a different block than what we wound up finalizing?

Is it worth checking the hash that was signed before we compute the union or do we not have to worry about this

Reply:
After talking this over with @kevjue, it does seem like there is an issue here.

Imagine I'm the block proposer, and I send you block A and everyone else block B. I then send you a COMMIT for block A. Everyone else agrees on B, and it eventually gets broadcast to you with 2F+1 signatures. You accept block B, and move on to a new sequence, copying the COMMIT I sent you for block A into your `ParentCommits`.

If you are the next block proposer, you will wind up proposing a block with parent signatures from both block A and block B, which will be invalid. 

To fix this, it seems that we have two options:
- Before copying `Commits` over to `ParentCommits`, check that the most recent block's hash matches those in the `Commit` messages.
- Before doing the union here, check that the block hashes match in the same manner.

I think we have a slight preference for option 1, all things being equal.

Reply:
This fix should be address this @asaj: https://github.com/celo-org/celo-blockchain/pull/538/commits/32ed4438e77603500a1b04c4c5654000c265fcc8

Reply:
This should be addressed in https://github.com/celo-org/celo-blockchain/pull/538/commits/b8627f90bc7114d21088b95729215c122b368849